### DISCUSSION TITLE
Building Angular app for component tests with esbuild or reusing already built code for devServer

### DISCUSSION THREAD

**Discussion by mirobo** [cypress-io/cypress#27034] [Questions and Help]
Before running component tests, the application itself is built with either the bundler 'vite' or 'webpack'. I don't see a way around this limitation at the moment.. so as an alternative I thought of this:

Is it possible to configure Cypress component testing in a way that it reuses an already built Angular application? We have multiple smaller Cypress projects in a monorepo but all these projects use the same Angular app. It would be nice if we can somehow tell Cypress to use an already existing build. The build output could be created with "ng build --watch".

But how could I tell Cypress that it doesn't need to build the app?

**Comment by portyu9** [comment]
Hi @mirobo 

You cannot simply point Cypress Component Testing at the output of `ng build --watch` and tell it that the application is already built.

Component Testing does not run the already-built application. Its dev server has to compile and serve the **Cypress spec/support files plus the component dependency graph** so Cypress can mount an individual Angular component.

For Angular, Cypress's currently supported built-in Component Testing bundler is still **Webpack**.

There is one important change since this question was posted: Cypress now supports a **fully custom `component.devServer` function**. That lets you integrate another build/dev-server implementation if it can compile the Cypress specs and components and return the port Cypress should connect to. So a custom integration is now possible, but an ordinary `ng build --watch` output directory by itself is not enough.

If the real goal is to build the Angular application only once and share that build between several Cypress projects, that fits **E2E testing** better: serve the built application once and point each Cypress project's `baseUrl` at it.

For true Component Testing, Cypress still needs a dev server capable of bundling the component test itself.

Please mark this answer as accepted if it helped, thank you.
